src/buffer_client.cc

```diff
@@ -73,14 +73,31 @@ BufferClient::BufferClient(const std::string& server_node_name)
           lookup_transform_service_full, true /* persistent */);
 }
 
+BufferClient::BufferClient(const std::string& server_node_name,
+                           const ros::Duration keepalive_period)
+    : BufferClient(server_node_name) {
+  keepalive_timer_ =
+      node_handle_.createTimer(keepalive_period, [&](const ::ros::TimerEvent&) {
+        if (!isConnected()) {
+          asyncReconnect(ros::Duration(-1));
+        }
+      });
+}
+
 BufferClient::~BufferClient() { node_handle_.shutdown(); }
 
 bool BufferClient::reconnect(ros::Duration timeout) {
-  std::lock_guard<std::mutex> guard(mutex_);
+  std::unique_lock<std::mutex> reconnection_guard(reconnection_mutex_,
+                                                  std::try_to_lock);
+  if (!reconnection_guard.owns_lock()) {
+    ROS_WARN("Already reconnecting to tf_service server.");
+    return false;
+  }
   if (!can_transform_client_.waitForExistence(timeout)) {
-    ROS_ERROR("Failed to connect to server.");
+    ROS_ERROR("Failed to connect to tf_service server.");
```

Comment:
Should we also print the full name of the service? Sometime it can help in resolving/understanding eventual problems with names.
`"Failed to connect to tf_service server at << can_transform_client_.getService() << "."`

Reply:
Yes, sounds good.

Reply:
Done, I put it in the log before waiting, so you can see it in case the wait takes long.